src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+			Arc::clone(&logger),
+		));
+
+		//// Step 11: Initialize the PeerManager
+		let onion_messenger: Arc<OnionMessenger> = Arc::new(OnionMessenger::new(
+			Arc::clone(&keys_manager),
+			Arc::clone(&keys_manager),
+			Arc::clone(&logger),
+			IgnoringMessageHandler {},
+		));
+		let ephemeral_bytes: [u8; 32] = rand::thread_rng().gen();
+		let lightning_msg_handler = MessageHandler {
+			chan_handler: Arc::clone(&channel_manager),
+			route_handler: Arc::clone(&gossip_sync),
+			onion_message_handler: onion_messenger,
+		};
+
+		let cur_time = SystemTime::now()
+			.duration_since(SystemTime::UNIX_EPOCH)
+			.expect("System time error: Clock may have gone backwards");
+		let peer_manager: Arc<PeerManager> = Arc::new(PeerManager::new(
+			lightning_msg_handler,
+			keys_manager.get_node_secret(Recipient::Node).unwrap(),
+			cur_time.as_secs().try_into().expect("System time error"),
+			&ephemeral_bytes,
+			Arc::clone(&logger),
+			IgnoringMessageHandler {},
+		));
+
+		// Step 13: Init payment info storage
+		// TODO: persist payment info to disk
+		let inbound_payments = Arc::new(Mutex::new(HashMap::new()));
+		let outbound_payments = Arc::new(Mutex::new(HashMap::new()));
+
+		// Step 14: Restore event handler from disk or create a new one.
+		let event_queue = if let Ok(mut f) =
+			fs::File::open(format!("{}/{}", ldk_data_dir, event::EVENTS_PERSISTENCE_KEY))
+		{
+			Arc::new(
+				EventQueue::read(&mut f, Arc::clone(&persister))
+					.expect("Failed to read event queue from disk."),
+			)
+		} else {
+			Arc::new(EventQueue::new(Arc::clone(&persister)))
+		};
+
+		let peer_store = if let Ok(mut f) =
+			fs::File::open(format!("{}/{}", ldk_data_dir, peer_store::PEER_INFO_PERSISTENCE_KEY))
+		{
+			Arc::new(
+				PeerInfoStorage::read(&mut f, Arc::clone(&persister))
+					.expect("Failed to read peer information from disk."),
+			)
+		} else {
+			Arc::new(PeerInfoStorage::new(Arc::clone(&persister)))
+		};
+
+		let running = RwLock::new(None);
+
+		Node {
+			running,
+			config,
+			wallet,
+			tx_sync,
+			event_queue,
+			channel_manager,
+			chain_monitor,
+			peer_manager,
+			keys_manager,
+			network_graph,
+			gossip_sync,
+			persister,
+			logger,
+			router,
+			scorer,
+			inbound_payments,
+			outbound_payments,
+			peer_store,
+		}
+	}
+}
+
+/// Wraps all objects that need to be preserved during the run time of [`Node`]. Will be dropped
+/// upon [`Node::stop()`].
+struct Runtime {
+	tokio_runtime: Arc<tokio::runtime::Runtime>,
+	_background_processor: BackgroundProcessor,
+	invoice_payer:
+		Arc<InvoicePayer<Arc<EventHandler<Arc<FilesystemPersister>, Arc<FilesystemLogger>>>>>,
+	stop_networking: Arc<AtomicBool>,
+	stop_wallet_sync: Arc<AtomicBool>,
+}
+
+/// The main interface object of the simplified API, wrapping the necessary LDK and BDK functionalities.
+///
+/// Needs to be initialized and instantiated through [`Builder::build`].
+pub struct Node {
+	running: RwLock<Option<Runtime>>,
```

Comment:
Thinking aloud here... Regarding runtime management, you could have both a `Node` and a `RunningNode`, where the latter owns the former and a `Runtime`. `Node::start()` could take a `Runtime` and return a `RunningNode`, which would have any methods needing a `Runtime`. `RunningNode::stop()` could give back the `Node`.

Haven't really thought it all the way through, but the general idea would be to lean on the type system as much as possible.

Reply:
> Mh, generally a nice idea to rethink the approach taken for the `Runtime`. However, I see two issues with the outlined alternative approach:
> 
> 1. Giving a `Runtime` to `start()` significantly complicates the interface, especially since `Runtime` on its own would have no meaning or utility to the user. So it would be just an additional step they have to look up, i.e., an inconvenience to the user.

True, though it need not be a parameter if `start` could just create it. I hadn't looked closely enough to realize it wasn't `pub`.

> 2. While Rust's move semantics would allow us to make sure only one of the objects can be used at a time, I don't think this would work in bindings, where, as mentioned above, we'd never return just a `Node`/`RunningNode`, but always an `Arc<Node>`/`Arc<RunningNode>`. Result would be that the users might hold on to the former and hence they'd have two objects lying around that semantically should be the same thing.

I see...

I was thinking a bit about bindings for the builders used in offers. Currently each method takes `self` and returns either `Self` or `Result<Self, SemanticError>`. This won't work for bindings as you mentioned. I'm considering making a wrapper for bindings with a similar bindings-compatible interface, which would delegate to the idiomatic Rust builder and manage that internally.

Not sure if it makes sense to try something similar here where the bindings-compatible struct could internally manage `Node` / `RunningNode`. Drawback I suppose would be that the bindings interface would be different from the Rust interface.

Reply:
To expand upon the idea, the bindings version would hold an internal enum with `Running(RunningNode)` and `Stopped(Node)` variants and have the full set of methods, erroring on the ones requiring a `RunningNode`.

Reply:
Yeah, the rust version wouldn't let you call any methods needing a runtime if the node is not running. Maybe not worth it if the interfaces will differ and if it won't help with ridding an `Option` from the runtime in the wallet.